### PR TITLE
Make rank return an Integer object

### PR DESCRIPTION
**Pull request by kryzar**
### :books: Description

The following raises an error:

```python
Fq = GF(3)
A = Fq['T']
K.<T> = Frac(A)
phi = DrinfeldModule(A, [T, 0, 1])
rank = phi.rank()
cat = phi.category()
psi = cat.random_object(rank)
```

```
...
TypeError: rank must be a positive integer
```

There are at least two possible fixes:
1. Make `rank` return an `Integer` object.
2. Make `cat.random_object` accept both `Integer`s and `int`s.

I suggest the former, and provide an implementation.

### :memo: Checklist

- [X] The title is concise, informative, and self-explanatory.
- [X] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [X] I have created tests covering the changes.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by DavidAyotte** [comment]
This seems like a good fix, but I think a better fix would be to make the degree method of an Ore polynomial return an Integer object:
```
sage: P.<t> = OrePolynomialRing(GF(5)['T'], GF(5)['T'].frobenius_endomorphism())
sage: type(t.degree())
<class 'int'>
```

**Comment by kryzar** [comment]
> This seems like a good fix, but I think a better fix would be to make the degree method of an Ore polynomial return an Integer object:
> 
> ```
> sage: P.<t> = OrePolynomialRing(GF(5)['T'], GF(5)['T'].frobenius_endomorphism())
> sage: type(t.degree())
> <class 'int'>
> ```

Agreed.

**Comment by mkoeppe** [comment]
The docstring linter complains:
```
./sage/rings/function_field/drinfeld_modules/drinfeld_module.py:1178:1: RST218 Literal block expected; none found.
```

**Comment by xcaruso** [comment]
> This seems like a good fix, but I think a better fix would be to make the degree method of an Ore polynomial return an Integer object:
> 
> ```
> sage: P.<t> = OrePolynomialRing(GF(5)['T'], GF(5)['T'].frobenius_endomorphism())
> sage: type(t.degree())
> <class 'int'>
> ```

I also agree. Antoine, are you going to implement this?

**Comment by kryzar** [comment]
> > This seems like a good fix, but I think a better fix would be to make the degree method of an Ore polynomial return an Integer object:
> > ```
> > sage: P.<t> = OrePolynomialRing(GF(5)['T'], GF(5)['T'].frobenius_endomorphism())
> > sage: type(t.degree())
> > <class 'int'>
> > ```
> 
> I also agree. Antoine, are you going to implement this?

I can!

**Comment by kryzar** [comment]
The degree of an Ore polynomial should now be an `Integer`. I hope I did everything that needed to be done!

You'll notice that some tests fail: they are related to multi polynomials, and are also present on the `develop` branch. I therefore ignored them.

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 1]
```diff
@@ -2513,8 +2513,25 @@ cdef class OrePolynomial_generic_dense(OrePolynomial):
 
             sage: S(0).degree()
             -1
+
+        TESTS:
+
+        We check that the degree is an ``Integer`` object (see #35519)::
+
```
Comment (DavidAyotte) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]:
```suggestion

```

There are spaces characters here and the linter complaints.

**Comment by DavidAyotte** [comment]
Thanks Antoine! It all looks good to me. I'll approve once the linter is fixed and the actions are done.

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 2]
```diff
@@ -2513,8 +2513,25 @@ cdef class OrePolynomial_generic_dense(OrePolynomial):
...
+        We check that the degree is an ``Integer`` object (see #35519)::
+            
+            sage: R.<t> = ZZ[]
+            sage: sigma = R.hom([t+1])
+            sage: S.<x> = R['x',sigma]
+            sage: a = x^2 + t*x^3 + t^2*x + 1
+            sage: isinstance(a.degree(), Integer)
+            True
+
+            ::
```
Comment (xcaruso) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]:
```suggestion
        ::
```

File: src/sage/rings/function_field/drinfeld_modules/drinfeld_module.py [hunk 3]
```diff
@@ -1173,8 +1173,20 @@ def rank(self):
...
+            sage: Fq = GF(25)
+            sage: A.<T> = Fq[]
+            sage: K.<z12> = Fq.extension(6)
+            sage: p_root = 2*z12^11 + 2*z12^10 + z12^9 + 3*z12^8 + z12^7 + 2*z12^5 + 2*z12^4 + 3*z12^3 + z12^2 + 2*z12
+            sage: phi = DrinfeldModule(A, [p_root, z12^3, z12^5])
+            sage: isinstance(phi.rank(), Integer)
+            True
         """
-        return self._gen.degree()
+        return Integer(self._gen.degree())
```
Comment (xcaruso) [inline src/sage/rings/function_field/drinfeld_modules/drinfeld_module.py]: I would say that this change is no longer needed.

Comment (kryzar) [reply src/sage/rings/function_field/drinfeld_modules/drinfeld_module.py]: Ok.

**Review by DavidAyotte — APPROVED** [review]
Looks good to me! Thanks!

**Review by xcaruso — APPROVED** [review]
Looks good to me as well.

**Comment by kryzar** [comment]
Thank you Volker.

**Pull request merged**